runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,477 @@
+    }
+
+    async fn read_at(&self, buf: &mut [u8], offset: u64) -> Result<(), Error> {
+        let current_len = self.len.load(Ordering::Relaxed);
+        if offset + buf.len() as u64 > current_len {
+            return Err(Error::BlobInsufficientLength);
+        }
+
+        let fd = types::Fd(self.fd.as_raw_fd());
+        let mut total_read = 0;
+
+        let mut io_sender = self.io_sender.clone();
+        while total_read < buf.len() {
+            // Figure out how much is left to read and where to read into
+            let remaining = &mut buf[total_read..];
+            let offset = offset + total_read as u64;
+
+            // Create an operation to do the read
+            let op = opcode::Read::new(fd, remaining.as_mut_ptr(), remaining.len() as _)
+                .offset(offset as _)
+                .build();
+
+            // Submit the operation
+            let (sender, receiver) = oneshot::channel();
+            io_sender
+                .send((op, sender))
+                .await
+                .map_err(|_| Error::ReadFailed)?;
+
+            // Wait for the result
+            let bytes_read = receiver.await.map_err(|_| Error::ReadFailed)?;
+
+            // If the return value is non-positive, it indicates an error.
+            let bytes_read: usize = bytes_read.try_into().map_err(|_| Error::ReadFailed)?;
+            if bytes_read == 0 {
+                // Got EOF before filling buffer.
+                return Err(Error::BlobInsufficientLength);
+            }
+
+            total_read += bytes_read;
+        }
+
+        Ok(())
+    }
+
+    async fn write_at(&self, buf: &[u8], offset: u64) -> Result<(), Error> {
+        let fd = types::Fd(self.fd.as_raw_fd());
+        let mut total_written = 0;
+
+        let mut io_sender = self.io_sender.clone();
+        while total_written < buf.len() {
+            // Figure out how much is left to write and where to write from
+            let remaining = &buf[total_written..];
+            let offset = offset + total_written as u64;
+
+            // Create an operation to do the write
+            let op = opcode::Write::new(fd, remaining.as_ptr(), remaining.len() as _)
+                .offset(offset as _)
+                .build();
+
+            // Submit the operation
+            let (sender, receiver) = oneshot::channel();
+            io_sender
+                .send((op, sender))
+                .await
+                .map_err(|_| Error::WriteFailed)?;
+
+            // Wait for the result
+            let return_value = receiver.await.map_err(|_| Error::WriteFailed)?;
+
+            // If the return value is non-positive, it indicates an error.
+            let bytes_written: usize = return_value.try_into().map_err(|_| Error::WriteFailed)?;
+            if bytes_written == 0 {
+                return Err(Error::WriteFailed);
+            }
+
+            total_written += bytes_written;
+        }
+
+        // Update the virtual file size
```

Comment:
nit: maybe we should remove `len()`? If we put this after the write has completed, this will cause some weird race conditions I think (and if we can't maintain the interface we should probably drop it)?

Maybe we can just return it during `open` and not maintain it on the `Blob` itself?